numpy/_core/tests/test_datetime.py

```diff
@@ -2552,6 +2558,80 @@ def test_limit_str_roundtrip(self, time_unit, sign):
         limit_via_str = np.datetime64(str(limit), time_unit)
         assert limit_via_str == limit
 
+    def test_datetime_hash(self):
+        nat1 = np.datetime64()
+        nat2 = np.datetime64()
+        assert nat1 is not nat2
+        assert nat1 != nat2
+        assert hash(nat1) != hash(nat2)
+
+        dt = np.datetime64(2348, 'W')  # 2015-01-01
+        for unit in ('Y', 'M', 'W', 'D', 'h', 'm', 's', 'ms', 'us'):
+            dt2 = np.datetime64(dt, unit)
+            _assert_equal_hash(dt, dt2)
+
+            dt3 = np.datetime64(int(dt2.astype(int)) + 1, unit)
+            assert hash(dt) != hash(dt3)  # doesn't collide
+
+        for unit in ('h', 'm', 's', 'ms', 'us'):
+            dt2 = np.datetime64(dt, unit)
+            pydt = dt2.astype(datetime.datetime)
+            assert isinstance(pydt, datetime.datetime)
+            _assert_equal_hash(pydt, dt2)
+
+        dt = np.datetime64(-102894, 'W')  # -002-01-01
+        for unit in ('Y', 'M', 'W', 'D', 'h', 'm', 's', 'ms', 'us'):
+            dt2 = np.datetime64(dt, unit)
+            _assert_equal_hash(dt, dt2)
+
+        dt = np.datetime64(3, 'ms')
```

Comment:
any reason not to start with "s" here?  in pandas we rely on the invariance working for ['s', 'ms', 'us', 'ns']

Reply:
Only because we can't convert from seconds to attoseconds without integer overflow.
I split this test into two ranges: 1. minutes ... femtoseconds (which should keep Pandas happy) and 2. nanoseconds ... attoseconds.